### PR TITLE
Input sliders and Bounds for CPU and RAM

### PR DESCRIPTION
**Pull request by sdess09**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Other (please describe):

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] If it's a backend change, tests for the changes have been added and `go test ./...` runs successfully from the root folder.
- [ ] If it's a frontend change, Prettier has been run
- [ ] Docs have been reviewed and added / updated if needed

### CODE REVIEW

File: dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx [hunk 1]
```diff
@@ -28,9 +42,133 @@ const WebTabs: React.FC<Props> = ({
...
+  const UPPER_BOUND = .75;
   const [currentTab, setCurrentTab] = React.useState<string>("main");
   const { currentCluster } = useContext(Context);
+  const context = useContext(Context);
+  const [nodeList, setNodeList] = useState([]);
+  const [maxCPU, setMaxCPU] = useState(2 * UPPER_BOUND); //default is set to a t3 medium 
+  const [maxRAM, setMaxRAM] = useState(4 * UPPER_BOUND); //default is set to a t3 medium
+  const [error, setError] = useState(false);
+
+  const awsInstanceLimits: InstanceTypes = {
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx:56]: add this as a const (AWS_INSTANCE_LIMITS) in utils.ts

File: dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx [hunk 2]
```diff
@@ -28,9 +42,133 @@ const WebTabs: React.FC<Props> = ({
...
+
+                  }
+                }
+              }
+            });
+
+            setMaxCPU(largestInstanceType.vCPUs * UPPER_BOUND);
+            setMaxRAM(largestInstanceType.RAM * UPPER_BOUND);
+
+            console.log(data);
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx:164]: remove

File: go.mod [hunk 3]
```diff
@@ -78,7 +78,7 @@ require (
 	github.com/honeycombio/otel-launcher-go v0.2.0
 	github.com/nats-io/nats.go v1.24.0
 	github.com/open-policy-agent/opa v0.44.0
-	github.com/porter-dev/api-contracts v0.0.75
+	github.com/porter-dev/api-contracts v0.0.78
```
Comment (ferozemohideen) [inline go.mod:81]: is this on purpose

Comment (sdess09) [reply go.mod:81]: yes the api-contracts was updated

File: dashboard/src/components/porter/InputSlider.tsx [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+  setValue: (value: number) => void;
+  disabled?: boolean;
+  disabledTooltip?: string;
+  color?: string;
+  width?: string;
+  step?: number;
+};
+
+const ValueLabelComponent: React.FC<any> = (props) => {
+  const { children, open, value } = props;
```
Comment (ferozemohideen) [inline dashboard/src/components/porter/InputSlider.tsx:25]: open is not used

File: dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx [hunk 5]
```diff
@@ -28,9 +30,81 @@ const WebTabs: React.FC<Props> = ({
   service,
   editService,
   setHeight,
+  chart,
 }) => {
+  const UPPER_BOUND = .75;
   const [currentTab, setCurrentTab] = React.useState<string>("main");
   const { currentCluster } = useContext(Context);
+  const context = useContext(Context);
+  const [nodeList, setNodeList] = useState([]);
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx:39]: why the nodeList and error variables?

File: dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx [hunk 6]
```diff
@@ -28,9 +30,81 @@ const WebTabs: React.FC<Props> = ({
...
+  const [error, setError] = useState(false);
+
+  useEffect(() => {
+    const { currentCluster, currentProject } = context;
+
+    if (!currentCluster || !currentProject) {
+      return;
+    }
+
+    const serviceName = service?.name;
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx:51]:
```suggestion
    const serviceName = service.name;
```

File: dashboard/src/main/home/app-dashboard/new-app-flow/tabs/utils.ts [hunk 7]
```diff
@@ -1,4 +1,66 @@
...
+        "nano": { "vCPU": 1, "Mem (GiB)": 0.5 },
+        "micro": { "vCPU": 1, "Mem (GiB)": 1 },
+        "small": { "vCPU": 1, "Mem (GiB)": 2 },
+        "medium": { "vCPU": 2, "Mem (GiB)": 4 },
+        "large": { "vCPU": 2, "Mem (GiB)": 8 },
+        "xlarge": { "vCPU": 4, "Mem (GiB)": 16 },
+        "2xlarge": { "vCPU": 8, "Mem (GiB)": 32 }
+    },
+    "c6i": {
+        "large": { "vCPU": 2, "Mem (GiB)": 4 },
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/tabs/utils.ts:46]: the keys should just be cpu and memory

**Review by ferozemohideen — CHANGES_REQUESTED** [review]
This is missing sliders for Job, Worker and Pre-deploy tabs. I would recommend you separate out the CPU/RAM sliders into its own component (using the InputSlider component you made) then reuse that parent component in each of WebTabs.tsx, WorkerTabs.tsx, JobTabs.tsx and ReleaseTabs.tsx (we should rename that to PreDeployTabs.tsx)

File: package-lock.json [hunk 8]
```diff
@@ -0,0 +1,1707 @@
+{
```
Comment (ferozemohideen) [inline package-lock.json:1]: the diff in this file is very large, i don't think you need to change all this?

File: dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx [hunk 9]
```diff
@@ -31,7 +34,74 @@ const ServiceContainer: React.FC<ServiceProps> = ({
   setExpandedJob,
 }) => {
   const [height, setHeight] = React.useState<Height>("auto");
+  const [nodeList, setNodeList] = useState([]);
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx:37]: nodelist isn't used

File: dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx [hunk 10]
```diff
@@ -31,7 +34,74 @@ const ServiceContainer: React.FC<ServiceProps> = ({
...
+  const UPPER_BOUND = .75;
 
+  const [maxCPU, setMaxCPU] = useState(2 * UPPER_BOUND); //default is set to a t3 medium 
+  const [maxRAM, setMaxRAM] = useState(4 * UPPER_BOUND); //default is set to a t3 medium
+  const [error, setError] = useState(false);
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx:42]: error isn't used

File: dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx [hunk 11]
```diff
@@ -31,7 +34,74 @@ const ServiceContainer: React.FC<ServiceProps> = ({
...
+  const context = useContext(Context);
+
+  useEffect(() => {
+    const { currentCluster, currentProject } = context;
+    if (!currentCluster || !currentProject) {
+      return;
+    }
+
+    const serviceName = service?.name;
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx:51]:
```suggestion
    const serviceName = service.name;
```

File: dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx [hunk 12]
```diff
@@ -31,7 +34,74 @@ const ServiceContainer: React.FC<ServiceProps> = ({
...
+    var instanceType = "";
+
+    //first check if there is a nodeSelector for the given application (Can be null)
+    if (chart?.config?.[serviceName + "-web"]?.nodeSelector?.["beta.kubernetes.io/instance-type"]) {
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx:55]: why "-web"?

File: dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx [hunk 13]
```diff
@@ -31,7 +34,74 @@ const ServiceContainer: React.FC<ServiceProps> = ({
...
+      instanceType = chart?.config?.[serviceName + "-web"]?.nodeSelector?.["beta.kubernetes.io/instance-type"]
+      const [instanceClass, instanceSize] = instanceType.split('.');
+      let currentInstance = AWS_INSTANCE_LIMITS[instanceClass][instanceSize];
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx:58]:
```suggestion
      const currentInstance = AWS_INSTANCE_LIMITS[instanceClass][instanceSize];
```

File: dashboard/src/main/home/app-dashboard/new-app-flow/tabs/ReleaseTabs.tsx [hunk 14]
```diff
@@ -42,27 +47,37 @@ const ReleaseTabs: React.FC<Props> = ({
...
-                    value={service.ram.value}
+                <InputSlider
+                    label="RAM: "
+                    unit="GiB"
+                    min={0}
+                    max={maxRAM}
+                    color={"#3a48ca"}
+                    value={(service.ram.value / 1024).toString()}
+                    setValue={(e) => {
+                        editService({ ...service, ram: { readOnly: false, value: e * 1024 } });
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/tabs/ReleaseTabs.tsx:77]: let's move these multipliers to consts in utils.ts

File: dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx [hunk 15]
```diff
@@ -31,7 +34,74 @@ const ServiceContainer: React.FC<ServiceProps> = ({
...
+
+            let largestInstanceType = {
+              vCPUs: 2,
+              RAM: 4,
+            };
+
+            data.forEach(node => {
+              if (node.labels['porter.run/workload-kind']) {
+                var instanceType: string = node.labels['beta.kubernetes.io/instance-type'];
+                const [instanceClass, instanceSize] = instanceType.split('.');
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx:85]: will this throw an error if instanceType is null?

File: dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx [hunk 16]
```diff
@@ -360,9 +372,11 @@ const WebTabs: React.FC<Props> = ({
             }
           />
           <Spacer y={1} />
-          <Input
-            label="Target CPU utilization (%)"
-            placeholder="ex: 50"
+          <InputSlider
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx:375]: any reason why the sliders for target utilization are a different width than the ones for ram/cpu? looks a little weird imo

File: package-lock.json [hunk 17]
```diff
@@ -0,0 +1,1707 @@
...
+  "name": "porter",
+  "lockfileVersion": 3,
+  "requires": true,
```
Comment (ferozemohideen) [inline package-lock.json:4]: can you delete this file then run npm i --legacy-peer-deps? I wanna see if that reduces the diff size

File: package.json [hunk 18]
```diff
@@ -0,0 +1,7 @@
+{
+  "dependencies": {
+    "@material-ui/core": "^4.12.4",
+    "@material-ui/lab": "^4.0.0-alpha.61",
+    "@openapitools/openapi-generator-cli": "^2.6.0"
+  }
+}
```
Comment (ferozemohideen) [inline package.json:7]: I don't think these needs to be here? I think they're already in dashboard/package.json. Also are you using the openapitools?

**Review by ferozemohideen — CHANGES_REQUESTED** [review]

File: dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx (hunk 9 above)
Comment (sdess09) [reply dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx:37]: removed

File: dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx (hunk 12 above)
Comment (sdess09) [reply dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx:55]: good catch

File: dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx (hunk 15 above)
Comment (sdess09) [reply dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx:85]: No in the case of our (node.labels['porter.run/workload-kind') will only exist if there is a instance type (in the case of our default cluster it doesn't exist and runs fine)

File: dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx (hunk 16 above)
Comment (sdess09) [reply dashboard/src/main/home/app-dashboard/new-app-flow/tabs/WebTabs.tsx:375]: width is now set to 300px for all

File: dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx [hunk 19]
```diff
@@ -32,6 +35,74 @@ const ServiceContainer: React.FC<ServiceProps> = ({
...
+    }
+    var instanceType = "";
+
+
+    if (service) {
+      const serviceName = service.name;
+
+      //first check if there is a nodeSelector for the given application (Can be null)
+      if (chart?.config?.[serviceName + service.type]?.nodeSelector?.["beta.kubernetes.io/instance-type"]) {
+        instanceType = chart?.config?.[serviceName + "-web"]?.nodeSelector?.["beta.kubernetes.io/instance-type"]
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx:57]: does this work for job/worker?

Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/new-app-flow/ServiceContainer.tsx:57]:
```suggestion
      if (chart?.config?.[`${serviceName}-${service.type}`]?.nodeSelector?.["beta.kubernetes.io/instance-type"]) {
        instanceType = chart?.config?.[`${serviceName}-${service.type}`]?.nodeSelector?.["beta.kubernetes.io/instance-type"]
```

**Review by ferozemohideen — APPROVED** [review]

**Pull request merged**